Add migration for mass relabeling relationship nodes

JustinAiken opened the pull request:
> So we're in the midst of updating the neo4j gem from 3.x to 6.x... One of the pain points was the way the relationship labels changed - `#old_way` to `NEW_WAY` (which is a change that makes sense and I'm glad you did!) Anyways, I'd written a helper that went through all the existing relationships and renamed them... [was wondering the best way to contribute it upstream]

- - me

> Hey @JustinAiken! That sounds like it might be a good fit for https://github.com/neo4jrb/neo4j/blob/master/lib/neo4j/migration.rb ?
You just need a class inheriting from `Neo4j::Migration` and the three methods in the public interface

- - @subvertallchris

Here's my first rough whack at fitting into a subclass of that... if this doesn't fit the project well I'll close and maybe create an outside gem, but if you like where it's heading I can clean it up, add more test cases, and do docs/changelog/etc

JustinAiken commented on spec/e2e/migration_spec.rb:
```diff
@@ -13,7 +13,7 @@ class ThirdRelClass; end
         include Neo4j::ActiveNode
         property :name
-        has_many :out, :songs, model_class: :Song, type: 'songs'
+        has_many :out, :songs, model_class: :'MigrationSpecs::Song', type: 'songs'
```
Before this change, doing `User.first.songs` gave me `ArgumentError: Could not find class: ::Song`

cheerfulstoic replied:
I actually did some refactoring of this just a couple of days ago so that the `MigrationSpecs` module isn't needed anymore.  I just merged [the PR](https://github.com/neo4jrb/neo4j/pull/1161) into master so you can try merging it into you branch if you like.  Otherwise we'll take care of it when this PR is done

JustinAiken replied:
Ah, nice refactor... rebased this branch off it, will handle the docs tomorrow :smile:

cheerfulstoic replied:
:metal:

subvertallchris commented:
This looks cool. Will it also move properties and clean up old data? We've had a few discussions about adding a method or methods to do this, too. See https://github.com/neo4jrb/neo4j/issues/1100 for one discussion. Maybe there are some reusable pieces that'll help us move towards that?

@cheerfulstoic, thoughts?

JustinAiken commented:
> Will it also move properties

No... I'm not sure how to make it do that.  I added a test case to demonstrate that properties do get stripped.  

> and clean up old data?

Yes! Added a test case for that

> We've had a few discussions about adding a method or methods to do this, too. See #1100 for one discussion. Maybe there are some reusable pieces that'll help us move towards that?

Looks like a related issue, but I don't see any way to combine the per-relationship-node-instance API with the mass update... I'm pretty unfamiliar with the code though, so if you do see anything reusable, let me know and I'll move it over!

cheerfulstoic commented on lib/neo4j/migration.rb:
```diff
@@ -123,5 +126,68 @@ def new_id_for(model)
+# Provide old and new label formats:
+# Allowed options are lower_hashtag, lower, or upper
+formats:\n  old: lower_hashtag\n  new: lower
+MESSAGE
```
Would be nice to see a full example of this `yml`.  Perhaps in the docs.  The docs are under the `docs` directory and they generate this site: http://neo4jrb.readthedocs.org/en/7.0.x/

cheerfulstoic replied:
Oh, and you can use `rake docs` and `rake docs:open` to test the docs locally

JustinAiken replied:
Just added the docs below... let me know if that's about what you had in mind!

cheerfulstoic commented on lib/neo4j/migration.rb:
```diff
@@ -123,5 +126,68 @@ def new_id_for(model)
+      private
+
+      def count(relationship)
+        Neo4j::Session.query(
```
Does this work?  I thought that you needed to do `Neo4j::Session.current.query`

JustinAiken replied:
The other migration task (`AddIdProperty`) does just `Neo4j::Session.query`, so I kept that...
In my real-world use on 6.1.10 there was no `.current` either..

cheerfulstoic replied:
Fair enough ;)

cheerfulstoic commented:
It looks like your code does move properties (the `SET r2 = r`).  But maybe that's not what it's doing because you seem to have a spec testing that it does strip the properties.

cheerfulstoic commented:
It seems fine with me.  I'm fine with it just being a straight Cypher query instead of re-using other code since I see this as mainly a way to help people upgrade versions

JustinAiken commented:
Screenshot of added doc page:

![screen](https://cloud.githubusercontent.com/assets/1568662/14002709/4be9935a-f113-11e5-8e99-c7d40dc2eca8.png)

cheerfulstoic commented:
LGTM!  @subvertallchris merge if it looks good to you?

subvertallchris commented:
I'll give it a close look again tomorrow!

On Thursday, March 24, 2016, Brian Underwood <notifications@github.com>
wrote:

> LGTM! @subvertallchris <https://github.com/subvertallchris> merge if it
> looks good to you?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/neo4jrb/neo4j/pull/1166#issuecomment-200660050>
>

subvertallchris commented:
It looks good but I have concerns about it not migrating properties. It's something I'd expect in a feature like this, especially since it's going to delete the old relationship afterwards. What about including properties to be migrated within the YML file?

subvertallchris commented:
Hey @JustinAiken, I'd love to get back into this if you're still interested. I'm happy to take a stab at adding the property migration if you want. What do you think?

JustinAiken commented:
Ah yes... I'm not sure what to do about migrating properties... All of our relationship nodes are just empty labels, so it never came up for us.  If you have an idea on how to rewrite the new relationships including the old properties, I'd love to hear it! 😄

subvertallchris commented:
Cool! I'll probably work off your branch and open a PR that covers it. Is
that ok?

On Monday, April 11, 2016, Justin Aiken <notifications@github.com> wrote:

> Ah yes... I'm not sure what to do about migrating properties... All of our
> relationship nodes are just empty labels, so it never came up for us. If
> you have an idea on how to rewrite the new relationships including the old
> properties, I'd love to hear it! 😄
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/neo4jrb/neo4j/pull/1166#issuecomment-208541762>
>

JustinAiken commented:
> Cool! I'll probably work off your branch and open a PR that covers it. Is
that ok?

Of course! :)

ProGM commented:
Closing here :)
